### ISSUE TITLE
Tree.sites iterator isn't reversable

### ISSUE THREAD

**Issue by jeromekelleher** [tskit-dev/tskit#2117]
It would be good to make the [Tree.sites](https://github.com/tskit-dev/tskit/blob/d421b4396a46472f74d8667a37413c188d1bddaf/python/tskit/trees.py#L1972) method reversable so that we can get at the sites backwards when running through the trees in reverse. 

You can always just to ``reversed(list(tree.sites()))`` at the moment, though.

While making the change, we should update the low-level API to return a numpy array of site IDs.

It's then a question of whether it's worth keeping the iterator behaviour or if we should just return a list of site objects. I guess there is potentially some performance creating Site objects lazily on demand, but most of the time we'll just be creating them one after another.

An option would be to create a ``SiteRange`` object which looks a bit like the [TreeIterator](https://github.com/tskit-dev/tskit/blob/d421b4396a46472f74d8667a37413c188d1bddaf/python/tskit/trees.py#L3547) but also supports ``__getitem__`` and creates site objects on demand.

Any thoughts?

**Comment by molpopgen** [comment]
Keeping things idiomatic by returning an iterator seems like a good idea to me. We could have a second function that returns an array if anyone wants to deal with the entire array at once?

**Comment by jeromekelleher** [comment]
Yeah, an array of ``site_ids`` would probably be quite handy. Probably the ``SiteRange`` implementation is the way to go then, it's easy enough to implement.
